src/ac_t_umbrella.c

```diff
@@ -0,0 +1,386 @@
+
+static void aTUMB_setupAction(UMBRELLA_ACTOR* umbrella, int action);
+
+static void aTUMB_actor_ct(ACTOR* actor, GAME* game) {
+  UMBRELLA_ACTOR* umbrella = (UMBRELLA_ACTOR*)actor;
+  aTUMB_setupAction(umbrella, umbrella->tools_class.work0);
+}
+
+static void aTUMB_OngenTrgStart_sub(ACTOR* actor, u16 id) {
+  sAdo_OngenTrgStart(id, &actor->world.position);
+}
+
+static void aTUMB_OngenTrgStart(ACTOR* actor, int idx) {
+  switch(idx) {
+    case 1:
+      aTUMB_OngenTrgStart_sub(actor, 0x139);
+      break;
+    case 2:
+      aTUMB_OngenTrgStart_sub(actor, 0x10E);
+      break;
+  }
+}
+
+static void aTUMB_calc_model_scale_sub(xyz_t* scale, UMBRELLA_ACTOR* umbrella, int idx) {
+  static f32 e_open_sect_table[6] = { 0.0f, 7.0f, 11.0f, 18.0f, 22.0f, 26.0f };
+  static f32 e_open_scale_table[12] = {
+    0.0f, 0.0f, 0.0f, 0.0f, 0.5f, 1.0f,0.5f, 1.0f, 1.0f, 1.0f, 1.0f, 1.0f
+  };
+
+  static f32 kasa_open_sect_table[4] = { 0.0f, 15.0f, 22.0f, 26.0f };
+  static f32 kasa_open_scale_table[8] = {
+    3.0f, 0.15f, 3.0f, 0.15f, 1.0f, 1.0f, 0.9, 1.0f
+  };
+
+  static UMBRELLA_DATA_TABLE open_data_table[2] = {
+    { 6, e_open_sect_table, e_open_scale_table },
+    { 4, kasa_open_sect_table, kasa_open_scale_table }
+  };
+
+  static f32 e_close_sect_table[6] = { 0.0f, 4.0f, 12.0f, 15.0f, 22.0f, 30.0f };
+  static f32 e_close_scale_table[12] = {
+    1.0f, 1.0f, 1.0f, 1.0f, 0.5f, 1.0f, 0.5f, 1.0f, 0.0f, 0.0f, 0.0f, 0.0f
+  };
+
+  static f32 kasa_close_sect_table[4] = { 0.0f, 4.0f, 12.0f, 30.0f };
+  static f32 kasa_close_scale_table[8] = {
+    1.0f, 1.0f, 1.2f, 1.0f, 3.0f, 0.15f, 3.0f, 0.15f
+  };
+
+  static UMBRELLA_DATA_TABLE close_data_table[2] = {
+    { 6, e_close_sect_table, e_close_scale_table },
+    { 4, kasa_close_sect_table, kasa_close_scale_table }
+  };
+
+  static UMBRELLA_DATA_TABLE* data_table[6] = {
+    open_data_table, open_data_table, close_data_table, close_data_table, open_data_table, NULL
+  };
+
+  f32 num1;
+  f32 num2;
+  f32 num3;
+  f32 num4;
+  f32 num5;
+
+  int scale_idx_offset;
+  f32* scale_table;
+  int sector_idx;
+  int frame;
+  int corresponding_sector;
+  UMBRELLA_DATA_TABLE *table;
+  f32* sect_table;
+
+  corresponding_sector = 0;
+  table = data_table[umbrella->action] + idx;
+  sector_idx = table->table_length - 1;
+  frame = (int)umbrella->frame;
+  sect_table = table->sect_table + 1;
+
+  for(sector_idx; sector_idx > 0; sector_idx--) {
+    if (frame <= (int)*sect_table) {
+      frame -= (int)sect_table[-1];
+      break;
+  }
+
+    sect_table++;
+    corresponding_sector++;
+  }
+
+  scale_idx_offset = corresponding_sector * 2;
+  scale_table = table->scale_table;
+  num2 = scale_table[scale_idx_offset];
+  num1 = (f32) frame / (sect_table[0] - sect_table[-1]);
+  num3 = scale_table[scale_idx_offset+ 1];
+  num4 = scale_table[scale_idx_offset + 3];
+  num5 = scale_table[scale_idx_offset + 2];
+
+  scale->x = num2 + num1 * (num5 - num2);
+  scale->y = num3 + num1 * (num4 - num3);
```

Comment:
I don't like that generic numX names are used here. num2 is the base scale for X, num3 is the base scale for Y, num1 is the modifier percentage, num5 is the max scale for X and num4 is the max scale for Y.